### PR TITLE
[UX] Friendly error message when mounting fails with non-empty mount path

### LINKED ISSUE #1895: Bucket mounting should print friendlier error messages
Accidentally, I tried to mount to a path already taken up by the OS:
```
file_mounts:
  /tmp:
    name: tmp-workdir-zongheng
    source: ~/tmp-workdir
```
This throws a long error message:
```
<launching>
sky.exceptions.CommandError: Command (cat <<-\EOF > ~/.sky/mount_72018.sh
#!/usr/bin/env bash
set -e

MOUNT_PATH=/tmp
MOUNT_BINARY=gcsfuse

# Check if path is already mounted
if grep -q $MOUNT_PATH /proc/mounts ; then
    echo "Path already mounted - unmounting..."
    fusermount -uz "$MOUNT_PATH"
    echo "Successfully unmounted $MOUNT_PATH."
fi

# Install MOUNT_BINARY if not already installed
if [ -x "$(command -v gcsfuse)" ] && gcsfuse --version | grep -q 0.42.3; then
  echo "$MOUNT_BINARY already installed. Proceeding..."
else
  echo "Installing $MOUNT_BINARY..."
  wget -nc https://github.com/GoogleCloudPlatform/gcsfuse/releases/download/v0.42.3/gcsfuse_0.42.3_amd64.deb -O /tmp/gcsfuse.deb && sudo dpkg --install /tmp/gcsfuse.deb
fi

# Check if mount path exists
if [ ! -d "$MOUNT_PATH" ]; then
  echo "Mount path $MOUNT_PATH does not exist. Creating..."
  sudo mkdir -p $MOUNT_PATH
  sudo chmod 777 $MOUNT_PATH
else
  # Check if mount path contains files
  if [ "$(ls -A $MOUNT_PATH)" ]; then
    echo "Mount path $MOUNT_PATH is not empty. Please make sure its empty."
    exit 1
  fi
fi
echo "Mounting $SOURCE_BUCKET to $MOUNT_PATH with $MOUNT_BINARY..."
gcsfuse -o allow_other --implicit-dirs --stat-cache-capacity 4096 --stat-cache-ttl 5s --type-cache-ttl 5s --rename-dir-limit 10000 tmp-workdir-zongheng /tmp
echo "Mounting done."
) && chmod +x ~/.sky/mount_72018.sh && bash ~/.sky/mount_72018.sh && rm ~/.sky/mount_72018.sh failed with return code 1.
Failed to run command before rsync ~/tmp-workdir -> /tmp. Ensure that the network is stable, then retry.
```
We should

1. Skip printing the entire script
2. Print out the error messages from the underlying tool (probably something like `/tmp` is taken up)

### PR DESCRIPTION
**Pull request by landscapepainter**
This resolves #1895 

When launching a remote VM, there is a chance of the mount path set by the user to be pre-occupied when the path is taken by the kernel (i.e. `/tmp` or `/sys`). In such case, `sky` raises `CommandError` showing what command caused the error. But in this case, the command is a heredoc which causes the error message to be very long and difficult to interpret. This PR fixes the issue by suppressing the long command raised with CommandError and raises a concise and straightforward message.

Before:
```
sky.exceptions.CommandError: Command (cat <<-\EOF > ~/.sky/mount_300531.sh
#!/usr/bin/env bash
set -e

MOUNT_PATH=/tmp
MOUNT_BINARY=gcsfuse

# Check if path is already mounted
if grep -q $MOUNT_PATH /proc/mounts ; then
    echo "Path already mounted - unmounting..."
    fusermount -uz "$MOUNT_PATH"
    echo "Successfully unmounted $MOUNT_PATH."
fi

# Install MOUNT_BINARY if not already installed
if [ -x "$(command -v gcsfuse)" ] && gcsfuse --version | grep -q 0.42.3; then
  echo "$MOUNT_BINARY already installed. Proceeding..."
else
  echo "Installing $MOUNT_BINARY..."
  wget -nc https://github.com/GoogleCloudPlatform/gcsfuse/releases/download/v0.42.3/gcsfuse_0.42.3_amd64.deb -O /tmp/gcsfuse.deb && sudo dpkg --install /tmp/gcsfuse.deb
fi

# Check if mount path exists
if [ ! -d "$MOUNT_PATH" ]; then
  echo "Mount path $MOUNT_PATH does not exist. Creating..."
  sudo mkdir -p $MOUNT_PATH
  sudo chmod 777 $MOUNT_PATH
else
  # Check if mount path contains files
  if [ "$(ls -A $MOUNT_PATH)" ]; then
    echo "Mount path $MOUNT_PATH is not empty. Please make sure its empty."
    exit 1
  fi
fi
echo "Mounting $SOURCE_BUCKET to $MOUNT_PATH with $MOUNT_BINARY..."
gcsfuse -o allow_other --implicit-dirs --stat-cache-capacity 4096 --stat-cache-ttl 5s --type-cache-ttl 5s --rename-dir-limit 10000 tmp-doyoung-testing-root /tmp
echo "Mounting done."
) && chmod +x ~/.sky/mount_300531.sh && bash ~/.sky/mount_300531.sh && rm ~/.sky/mount_300531.sh failed with return code 1.
Failed to run command before rsync ~/tmp-workdir -> /tmp. Ensure that the network is stable, then retry.
```

After:
```
sky.exceptions.StorageMountPathError: Mount path /tmp is non-empty. /tmp may have been already taken by the Kernel. Please set the mount path to another name.
```

Tested (run the relevant ones):

- [x] Launching a single node cluster with pre-occupied directory, `/tmp`, set as mount path
- [x] Launching a multi-node cluster with pre-occupied directory, `/tmp`, set as mount path

### CODE REVIEW

File: sky/backends/cloud_vm_ray_backend.py [hunk 1]
```diff
@@ -3558,15 +3558,27 @@ def _execute_storage_mounts(self, handle: CloudVmRayResourceHandle,
...
+                )
+            except exceptions.CommandError as e:
+                if e.returncode == exceptions.MOUNT_PATH_NON_EMPTY_CODE:
+                    mount_path = (f'{colorama.Fore.RED}'
+                                  f'{colorama.Style.BRIGHT}{dst}'
+                                  f'{colorama.Style.RESET_ALL}')
+                    error_msg = (f'Mount path {mount_path} is non-empty.'
+                                 f' {mount_path} may have been already '
+                                 f'taken by the Kernel. Please set the '
+                                 f'mount path to another name.')
```
Comment (romilbhardwaj) [inline sky/backends/cloud_vm_ray_backend.py:3579]:
```suggestion
                    error_msg = (f'Mount path {mount_path} is non-empty.'
                                 f' {mount_path} may be a standard unix '
                                 f' path or may contain files from a previous task. To fix, change the '
                                 f'mount path to an empty or non-existent path.')
```

**Review by romilbhardwaj — COMMENTED** [review]
Thanks @landscapepainter! Left some comments.

File: sky/backends/cloud_vm_ray_backend.py [hunk 2]
```diff
@@ -3558,15 +3558,27 @@ def _execute_storage_mounts(self, handle: CloudVmRayResourceHandle,
...
+                    raise exceptions.StorageMountPathError(error_msg) from None
```
[the remark this answers is not in the document]
Comment (landscapepainter) [reply sky/backends/cloud_vm_ray_backend.py:3581]: With the current implementation, when `e.returncode != exceptions.MOUNT_PATH_NON_EMPTY_CODE` (manually setting the exit code of the heredoc to be `99` and `MOUNT_PATH_NON_EMPTY_CODE` to be `42`) and attempting to mount to a non-empty directory, it does surface the error message from the heredoc without displaying the entire heredoc : `E 05-05 03:11:59 subprocess_utils.py:70] Mount path /tmp is not empty. Please make sure its empty.`

Does it still need the `|| echo` implementation in `mounting_utils.py`? If so, what error should it be catching? The only error with an exit code in the heredoc from `mounting_utils.py` is when the `$MOUNT_PATH` is non-empty, and in this case, it is hard coded that it exits with `42` and echos a error message as well: `E 05-05 03:11:59 subprocess_utils.py:70] Mount path /tmp is not empty. Please make sure its empty.`

Or perhaps, I misunderstood what you meant. Would you mind elaborating a bit more if this is the case?

**Comment by landscapepainter** [comment]
@romilbhardwaj Ready for another look :)

File: sky/exceptions.py [hunk 3]
```diff
@@ -164,6 +165,12 @@ class StorageModeError(StorageSpecError):
     pass
 
 
+class StorageMountPathError(StorageSpecError):
+    # Error raised when bucket is successfully initialized, but mounting fails
+    # due to non-empty mount path
+    pass
```
Comment (concretevitamin) [inline sky/exceptions.py:171]: Nit: there's an issue from way before saying storage exceptions are too complex. Maybe we shouldn't add another one. How about a simple RuntimeError for now? When we have enough API usage in the wild, we can use that to decide how to design exceptions.

File: sky/data/mounting_utils.py [hunk 4]
```diff
@@ -60,7 +60,7 @@ def get_mounting_command(
           # Check if mount path contains files
           if [ "$(ls -A $MOUNT_PATH)" ]; then
             echo "Mount path $MOUNT_PATH is not empty. Please make sure its empty."
```
Comment (concretevitamin) [inline sky/data/mounting_utils.py:62]:
```suggestion
            echo "Mount path $MOUNT_PATH is not empty. Please mount to another path or remove it first."
```
(Suggest a fix for common scenarios first; e.g., if people accidentally mount to /tmp, the first fix should be use another path)

--- 

NOTE: I find it confusing that we allow mounting to a `/empty-dir` without files. The convention in other tools seems to be if the path exists, regardless of whether files exist, then back off.

File: sky/exceptions.py [hunk 5]
```diff
@@ -10,6 +10,7 @@
 KEYBOARD_INTERRUPT_CODE = 130
 SIGTSTP_CODE = 146
 RSYNC_FILE_NOT_FOUND_CODE = 23
+MOUNT_PATH_NON_EMPTY_CODE = 42
```
Comment (concretevitamin) [inline sky/exceptions.py:13]:
```suggestion
MOUNT_PATH_NON_EMPTY_CODE = 42  # Arbitrarily chosen.
```

**Review by concretevitamin — COMMENTED** [review]
Quick look at the UX.

File: sky/data/mounting_utils.py (hunk 4 above)
Comment (landscapepainter) [reply sky/data/mounting_utils.py:62]: Updated the comment.

What kind of potential issues can be arised by allowing to mount on existing directories?

**Comment by landscapepainter** [comment]
@concretevitamin @romilbhardwaj Ready for another look! Thanks!

**Review by romilbhardwaj — COMMENTED** [review]
Thanks @landscapepainter! Left some comments. Tried it out and the error is much cleaner now. It also prints the stdout/stderr of the mounting script, which I think is helpful for debugging.

**Comment by landscapepainter** [comment]
@romilbhardwaj ready for another look!

File: sky/data/mounting_utils.py [hunk 6]
```diff
@@ -59,8 +59,8 @@ def get_mounting_command(
         else
           # Check if mount path contains files
           if [ "$(ls -A $MOUNT_PATH)" ]; then
-            echo "Mount path $MOUNT_PATH is not empty. Please make sure its empty."
-            exit 1
+            echo "Mount path $MOUNT_PATH is not empty. Please mount to another path or remove it first."
+            exit 42
```
[the remark this answers is not in the document]
Comment (landscapepainter) [reply sky/data/mounting_utils.py]: Resolved!

**Review by romilbhardwaj — APPROVED** [review]
Great work @landscapepainter, thanks!

**Comment by concretevitamin** [comment]
Works nicely @landscapepainter - one UX observation, with the `/tmp: ...` repro:

```
...
I 05-20 08:56:40 backend_utils.py:1220] Mounting (to 1 node): ~/tmp-workdir -> /tmp
⠼ MountingE 05-20 08:56:44 subprocess_utils.py:70] bash: warning: here-document at line 37 delimited by end-of-file (wanted `EOF')
E 05-20 08:56:44 subprocess_utils.py:70] gcsfuse version 0.41.9 (Go version go1.18.4)
E 05-20 08:56:44 subprocess_utils.py:70] Installing gcsfuse...
...
E 05-20 08:56:44 subprocess_utils.py:70] Setting up gcsfuse (0.42.3) ...
E 05-20 08:56:44 subprocess_utils.py:70] Mount path /tmp is not empty. Please mount to another path or remove it first.
E 05-20 08:56:44 subprocess_utils.py:70] Warning: Permanently added '35.226.87.123' (ED25519) to the list of known hosts.
```
- Is the `here-document at line 37 delimited by end-of-file (wanted `EOF')` warning expected? Somewhere we should do write string EOF?
- Minor: The install log of gcsfuse (many lines) is printed with E, error. That seems a bit unexpected as really only the second to last line is erroring. If it's too hard to fix we can certainly drop it.

**Comment by landscapepainter** [comment]
@concretevitamin Thanks for bringing up those cases:
`EOF` seems to be a known issue
Screenshot 2023-05-20 at 5 50 37 PM
I tried adding string `'EOF'` at the end of here-doc and as part of `command` after `{script}`, but neither fixed.

For the gcfuse install log, `subprocess_utils.py/handle_returncode()` outputs `stderr + stdout` when the `returncode != 0`. And the install log is part of the `stdout` from `backend_utils.py/parallel_data_transfer_to_nodes()`. By the way, the `EOF` error is also part of `stdout` as well. So I think it's purposely designed to show `stdout` and `stderr` to help debug when `returncode != 0`

**Comment by concretevitamin** [comment]
> outputs `stderr + stdout` when the `returncode != 0`

Nit: Maybe we can logger.info() stdout and logger.error() stderr in this case?

**Pull request merged**